### PR TITLE
2392 broadcast similarity

### PR DESCRIPTION
**Pull request by domangi**

### CODE REVIEW

File: backend/app/models/similarity.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class Similarity < ApplicationRecord
+  belongs_to :broadcast1, foreign_key: 'broadcast1_id', class_name: 'Broadcast'
+  belongs_to :broadcast2, foreign_key: 'broadcast2_id', class_name: 'Broadcast'
+
+  def self.compute_all
+    Similarity.destroy_all
+    Broadcast.all.each do |broadcast1|
```
Comment (roschaefer) [inline backend/app/models/similarity.rb]: use `find_each`

File: backend/app/models/similarity.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def self.compute(broadcast1, broadcast2)
+    similarity = Similarity.new(broadcast1: broadcast1, broadcast2: broadcast2)
+    similarity.compute_jaccard
+    similarity
+  end
+
+  def compute_jaccard
+    supporters1 = broadcast1.impressions.where(response: 1).pluck(:user_id)
```
Comment (roschaefer) [inline backend/app/models/similarity.rb]: same as `broadcast1.impressions.positive.pluck(:user_id)`

**Comment by roschaefer** [comment]
I like Jaccard similarity for its damn simple implementation. However, all the database read/write operations are very expensive. It takes 9 minutes and 13 seconds on my machine to compute jaccard similarity on the production database. I would suggest to read the set of supporters for each broadcast once and keep that in memory. Then we can have the iteration through each pair of broadcast.

**Comment by roschaefer** [comment]
Btw.
```
irb(main):015:0> Similarity.count
   (6.1ms)  SELECT COUNT(*) FROM "similarities"
=> 44476
irb(main):016:0> Broadcast.count
   (0.7ms)  SELECT COUNT(*) FROM "broadcasts"
=> 485
irb(main):017:0> 485 * 485
=> 235225
irb(main):019:0> 44476 /  235225.0
=> 0.18907854182166012
```

**Comment by roschaefer** [comment]
And I would suggest to store `impressions.positive.count` for each broadcast in the column. Why?

```ruby
puts Similarity.order(value: :desc).take(20).map{|s| "Jaccard index #{s.value} for #{s.broadcast1.title}(#{s.broadcast1.impressions.positive.count}) and #{s.broadcast2.title}(#{s.broadcast2.impressions.positive.count})" }
"
Jaccard index 1.0 for Doppelkopf(1) and SWR 2 Aula(1)
Jaccard index 1.0 for Exakt - Die Story(1) and Asül für Alle(1)
Jaccard index 1.0 for Am Puls Deutschlands(1) and Asül für Alle(1)
Jaccard index 1.0 for Heldt(1) and Herzensbrecher - Vater von vier Söhnen(1)
Jaccard index 1.0 for Mako(1) and FritzFressefreiheit(1)
Jaccard index 1.0 for Exakt - Die Story(1) and Am Puls Deutschlands(1)
Jaccard index 1.0 for hessentipp(1) and Rhein-Main und Südhessen(1)
Jaccard index 1.0 for Filter(1) and Netzmagazin(1)
Jaccard index 1.0 for Exakt - Die Story(1) and Arte Jazz(1)
Jaccard index 1.0 for Hochfahren(1) and Plattenbau(1)
Jaccard index 1.0 for Asül für Alle(1) and Arte Jazz(1)
Jaccard index 1.0 for Radio-Visite(2) and MDR Hauptsache gesund(2)
Jaccard index 1.0 for Resonanzen(1) and TonArt(1)
Jaccard index 1.0 for Ein guter Grund zu feiern(1) and Willkommen bei Carmen Nebel(1)
Jaccard index 0.666666666666667 for Scala(2) and Theodor(3)
Jaccard index 0.666666666666667 for mehr/wert(2) and Bernama Kurdî(3)
Jaccard index 0.666666666666667 for defacto(3) and KiRaKa Klicker(2)
Jaccard index 0.666666666666667 for mehr/wert(2) and Tipps zu Gesundheit, Freizeit und Ernährung(3)
Jaccard index 0.666666666666667 for KiRaKa Klicker(2) and Mittagsmagazin(3)
"```

I would like to normalize the similarity somehow to take into account the number of total impressions per broadcast.

**Comment by roschaefer** [comment]
OK, I improved performance quite a lot. @domangi could you think of a good normalization?

I don't want to have the strongest similarities between two broadcasts that have only one supporter each. If and only if two broadcasts have many supporters and also share a lot of them - then they are similar to each other.

**Comment by domangi** [comment]
Yes, I knew the compute_all method was not very efficient and that it needed a performance tuning. Great job @roschaefer ! 

About the normalization I would not change the similarity score, because it makes sense that 2 broadcasts that share exactly the same supporters have 100% similarity, or in other words, it would make less sense giving them a smaller similarity score.

But I got your point, 2 broadcasts that are only supported by 1 person should not be considered similar, because this is not really enough information.
A solution could be to use a threshold: a broadcast to be considered in the similarity computation has to have at least N supporters, where N could be a manually configured parameter or could be calculated dynamically, e. g. greater than the median value or greater than 10% of the maximum value of supporters, ...
All broadcasts that are below the threshold will have no similarity with anyone, because they do not have enough supporters to allow computing a meaningful similarity score.

**Comment by roschaefer** [comment]
Hey @domangi apparently, the accuracy problem we're facing here is a well researched topic called [cold start problem](https://en.wikipedia.org/wiki/Cold_start). I did some research on the internet and there are numerous versions and modifications of existing similarity measures, including jaccard similarity. I also played with your suggestions of a general minimal number of supporters per broadcast. We get much better results for a threshold of e.g. `5`. I say, it's enough for now and if we want to improve we can modify the similarity measure later on. What do you think?

**Pull request merged**